File: ParkingSystem.java
import java.util.*;
import java.time.LocalDateTime;
import java.time.temporal.ChronoUnit;

public class Main {
    public static void main(String[] args) {
      System.out.println("Design a parking system");
  }
}

// vehicle to park in parking lot.
// Entites : Vehicle, ParkingTicket, ParkingLot, ParkingFloor


// Vehicle entity
abstract class Vehicle{
  
  private String licenseNumber;
  protected VehicleType type;
  
  public Vehicle(String licenseNumber, VehicleType type){
    this.licenseNumber = licenseNumber;
    this.type = type;
  }
}

public enum VehicleType{
  LARGE, MEDIUM, SMALL
}

public class Car extends Vehicle{
  
   public Car(String licenseNumber, VehicleType type) {
        super(licenseNumber, type);
        // Add additional attributes specific to Car if needed
    }
}

public class Bike extends Vehicle{
  
      public Bike(String licenseNumber, VehicleType type) {
        super(licenseNumber, type);
        // Add additional attributes specific to Bike if needed
    }
}

// Ticket entity
public class ParkingTicket {
  
  private String parkingId;
  private VehicleType type;
  private LocalDateTime issuedAt;
  private boolean isPaid;
  private double totalFee;
  private LocalDateTime paidAt;
  
  public ParkingTicket(String parkingId, VehicleType type){
    this.parkingId = parkingId;
    this.type = type;
    this.issuedAt = LocalDateTime.now();
    this.isPaid = false;
  }
  
  public void markPaid(double totalFee){
    this.totalFee = totalFee;
    this.paidAt = LocalDateTime.now();
    this.isPaid = true;
  }
}

abstract class TicketCostCalculator{
  
    protected static double getVehicleHourlyRateByVehicleType(VehicleType type){
    switch(type){
      case SMALL:
        return 20.00;
      case MEDIUM:
        return 50.00;
      case LARGE:
        return 100.00;
      default:
        throw new IllegalArgumentException("Unknown vehicle type");
    }
  }
    
  protected static double totalCost(LocalDateTime entryTime, LocalDateTime exitTime, 
  VehicleType type ){
    double hourlyRate = getVehicleHourlyRateByVehicleType(type);
    long totalParkingTime = getTotalHours(entryTime, exitTime);
    return totalParkingTime * hourlyRate;
  }  
  private static long getTotalHours(LocalDateTime entryTime, LocalDateTime exitTime){
    return ChronoUnit.HOURS.between(entryTime, exitTime);
  } 
}

// parking lot entity
public class ParkingLot{
  
  private List<ParkingFloor> parkingFloors;
  private List<ParkingTicket> issuedTickets;
  
  public ParkingLot(){
    this.parkingFloors = new ArrayList<>();
    this.issuedTickets = new ArrayList();
  }
  
  public ParkingSpot findAvailableSpot(VehicleType type){
    // implement logic to find available spot by vehicle type
    return null;
  }
  
  public ParkingTicket issueParkingTicket(Vehicle vehicle){
    ParkingSpot spot = findAvailableSpot(vehicle.type);
    if(spot!=null){
      spot.occupySpot();
      ParkingTicket ticket = new ParkingTicket(generateTicketId(), vehicle.type);
      issuedTickets.add(ticket);
      return ticket;
    }
    return null;
  }
  
  public double processPayment(LocalDateTime entryTime, LocalDateTime exitTime, VehicleType type){
    return TicketCostCalculator.totalCost(entryTime, exitTime, type);
  }
  
   private String generateTicketId(){
    return "TICKET_"+ System.currentTimeMillis();
  } 
}

public class ParkingFloor{
  
  private int floorNo;
  private long totalParkingCapacity;
  
  public ParkingFloor(int floorNo, long totalParkingCapacity){
    this.floorNo = floorNo;
    this.totalParkingCapacity = totalParkingCapacity;
  }
  
  public void setTotalCapacityByVehicleType(VehicleType type){
    // implement logic
  }
  public long getTotalAvailableParkingSpotByVehicleType(VehicleType type){
    
    // implement logic
    
    return 0L;
  } 
}

public class ParkingSpot{
  
  private String id;
  private VehicleType type;
  private boolean isOccupied;
  
  public ParkingSpot(String id, VehicleType type){
    this.id = id;
    this.type = type;
    this.isOccupied = false;
  }
  
  public void occupySpot(){
    this.isOccupied = true;
  }
  
  public void vacantSpot(){
    this.isOccupied = false;
  }
}

File: ParkingSystem.py
from enum import Enum
from datetime import datetime, timedelta

class VehicleType(Enum):
    SMALL = 1
    MEDIUM = 2
    LARGE = 3

class Vehicle:
    def __init__(self, license_number, type):
        self.license_number = license_number
        self.type = type

class Car(Vehicle):
    def __init__(self, license_number, type):
        super().__init__(license_number, type)
        # Add additional attributes specific to Car if needed

class Bike(Vehicle):
    def __init__(self, license_number, type):
        super().__init__(license_number, type)
        # Add additional attributes specific to Bike if needed

class ParkingTicket:
    def __init__(self, parking_id, type):
        self.parking_id = parking_id
        self.type = type
        self.issued_at = datetime.now()
        self.is_paid = False
        self.total_fee = 0.0
        self.paid_at = None

    def mark_paid(self, total_fee):
        self.total_fee = total_fee
        self.paid_at = datetime.now()
        self.is_paid = True

class TicketCostCalculator:
    @staticmethod
    def get_vehicle_hourly_rate_by_vehicle_type(type):
        if type == VehicleType.SMALL:
            return 20.00
        elif type == VehicleType.MEDIUM:
            return 50.00
        elif type == VehicleType.LARGE:
            return 100.00
        else:
            raise ValueError("Unknown vehicle type")

    @staticmethod
    def total_cost(entry_time, exit_time, type):
        hourly_rate = TicketCostCalculator.get_vehicle_hourly_rate_by_vehicle_type(type)
        total_parking_time = (exit_time - entry_time).total_seconds() / 3600
        return total_parking_time * hourly_rate

class ParkingLot:
    def __init__(self):
        self.parking_floors = []
        self.issued_tickets = []

    def find_available_spot(self, type):
        # implement logic to find available spot by vehicle type
        return None

    def issue_parking_ticket(self, vehicle):
        spot = self.find_available_spot(vehicle.type)
        if spot:
            spot.occupy_spot()
            ticket = ParkingTicket(self.generate_ticket_id(), vehicle.type)
            self.issued_tickets.append(ticket)
            return ticket
        return None

    def process_payment(self, entry_time, exit_time, type):
        return TicketCostCalculator.total_cost(entry_time, exit_time, type)

    def generate_ticket_id(self):
        return "TICKET_" + str(int(datetime.now().timestamp() * 1000))

class ParkingFloor:
    def __init__(self, floor_no, total_parking_capacity):
        self.floor_no = floor_no
        self.total_parking_capacity = total_parking_capacity

    def set_total_capacity_by_vehicle_type(self, type):
        # implement logic
        pass

    def get_total_available_parking_spot_by_vehicle_type(self, type):
        # implement logic
        return 0

class ParkingSpot:
    def __init__(self, id, type):
        self.id = id
        self.type = type
        self.is_occupied = False

    def occupy_spot(self):
        self.is_occupied = True

    def vacant_spot(self):
        self.is_occupied = False
